README.md
# Ephesoft.Transact.DSC
This powershell module can be used by running Import-Module in a PowerShell console

## Contributing
Please see the contributing.md file

# Module Usage
## Installing Ephesoft Transact with the Module
````powershell
      TransactSetup 'InstallTransact'
      {
        DatabaseType = 'SQL'
        DBCredential = $MyDBCred
        Ensure = 'Present'
        InstallerPath = 'C:\Path\To\Uncompressed\Installers'
        Version = '2020.1'
        LicenseFilePath = 'C:\Path\To\ephesoft.lic'
        SQLAuthTye = 'SQL'
      }
````

## Parameters

| Name                      | Required                                   | Description                                               |
| :------------------------ | :----------------------------------------- | :-------------------------------------------------------- |
| Database Type             | Yes                                        | The type of database to use ('MariaDB', 'SQL', Oracle)    |
| DBCredential              | Yes - Except if using SQL with Windows Auth| PS credential object containing database credentials      |
| Ensure                    | Yes                                        | 'Present' or 'Absent' - Installs/Removes Ephesoft Transact|
| LicenseFilePath           | Yes                                        | Full file path to the license file to be installed        |
| SQLAuthType               | Yes                                        | Specifies the SQL authentication to use ('Windows', SQL') |
| Version                   | Yes                                        | The version of Ephesoft Transct to Install (i.e. 2020.1)  |

## Limitations
* The module currently only supports local databases
* The module currently only supports Ephesoft Transact Versions 2019.1 or higher

CHANGELOG.md
#ChangeLog
Version changes are logged here using semantic versioning (https://semver.org/)

## Deprecated features
*These will be removed in the next major release*
 - None

0.2.1
-----
 - Fix bug during installation when space is present in installer path
 - Fixed bug with licensing

0.2.0
-----
 - Reconfigure nested modules to support PowerShell 7

0.1.0
-----
 - Initial Commit

- - - - -
Check the [Markdown Syntax Guide](https://confluence.atlassian.com/bitbucketserver/markdown-syntax-guide-776639995.html) for basic syntax.

tests/ScriptAnalyzer.tests.ps1
Describe 'ScriptAnalyzer' {
	Context 'Validating ScriptAnalyzer installation' {
		It 'Checking Invoke-ScriptAnalyzer exists.' {
			{ Get-Command Invoke-ScriptAnalyzer -ErrorAction Stop } | Should Not Throw
		}
	}
}

Describe 'ScriptAnalyzer issues found' {

  # Excluding file 'MOFGnerator' because script analyzer in Bitbucket pipelines cannot parse the DSC configuration
  $ExcludedFiles = @('tests', 'examples')
  $ScriptAnalyzerSettings = @{

    <#
      PSDSCDscExamplesPresent and PSDSCDscTestsPresent excluded for running in Bitbucket pipelines
      The Ephesoft.Transact.DSC class based module cannot be imported in Linux so these tests will
      not succeed even though the examples and tests are present.
      Powershell DSC for Linux page: https://github.com/microsoft/PowerShell-DSC-for-Linux

      PSAvoidUsingCmdletAliases excluded due to 'Package' being overloaded by the DSC resource name
      'Package' and the alias for 'Get-Package'.  The Package DSC resource is referenced
      Transact.SQL.Example.ps1.
    #>
    ExcludeRules = @('PSDSCDscExamplesPresent','PSDSCDscTestsPresent','PSAvoidUsingCmdletAliases')
  }
	$results = Get-ChildItem .\* -Exclude $ExcludedFiles | Invoke-ScriptAnalyzer -Settings $ScriptAnalyzerSettings
	$scripts = $results.ScriptName | Get-Unique

	Context 'Checking results' {
		It 'Should have no issues' {
			$results.count | Should Be 0
		}
	}

	foreach ($script in $scripts) {
		Context $script {
			$issues = $results | Where-Object {$_.ScriptName -eq $script}

			foreach ($issue in $issues) {
				It "On line: $($issue.Line) - $($issue.Message)" {
					$true | Should Be $False
				}
			}
		}
	}
}

tests/TransactSetup.Get.Tests.ps1
using module '.\TransactSetup.TestHelper.psm1'

# Create object for global use
$Instance = New-TransactSetupObject

Describe "TransactSetup.Get()" {

  Context "Transact service exists" {
    Mock Get-Service -ModuleName TransactSetup {$True}

    $Data = $Instance.Get()
    It "Returns Ensure as Present" {
      $Data.Ensure | Should -Be 'Present'
    }
  }

  Context "Transact service does not exist" {
    Mock Get-Service -ModuleName TransactSetup {}
    $Data = $Instance.Get()

    It "Returns Ensure as Absent" {
      $Data.Ensure | Should -Be 'Absent'
    }
  }
}

tests/TransactSetup.Set.Tests.ps1
using module '.\TransactSetup.TestHelper.psm1'

# Set Global Mocks
Mock Get-ChildItem -ModuleName 'TransactSetup' {[pscustomobject]@{FullName = 'C:\Installer\Transact\Installer.msi'}}
Mock Start-Process -ModuleName 'TransactSetup' {}
Mock Set-Content -ModuleName 'TransactSetup' {}

# Set Global Variables
$InstallerPath = '.\tests\resources'
$InstallerMSI = 'C:\Installer\Transact\Installer.msi'
$Params = @{
  MethodsToMock = @('PrepareConfigProperties','Fix2020CronJobIssue','SetHostWebProperties','SetTransactLicense','StartService','DemandServiceExists')
  InstallerPath = $InstallerPath
}

# Create object for global use
$Instance = New-TransactSetupObject @Params

Describe "TransactSetup.Set()" {

  Context "Ensure is present when service does not exist" {
    $Instance = $Instance | Add-Member -MemberType ScriptMethod -Name TestServiceExists -Value { $false } -Force -PassThru

    # Run Test
    $Instance.Set()

    It "Passes Correct Installer Path To Start-Process" {
      Assert-MockCalled Start-Process -ModuleName 'TransactSetup' -ParameterFilter {
        $FilePath -eq $InstallerMSI
      }
    }

    It "Passes Correct Config File Path To Start-Process" {
      Assert-MockCalled Start-Process -ModuleName 'TransactSetup' -Exactly 1 -ParameterFilter {
        $ArgumentList -split(' ') -contains "USERINPUTSPATH=`"$InstallerPath\config.properties`""
      }
    }
  }

  Context "Ensure is Present When Service Already Exists" {
    Mock Get-Service -ModuleName 'TransactSetup' {$true}
    Mock Start-Service -ModuleName 'TransactSetup' {}

    # Run Test
    $Instance = $Instance | Add-Member -MemberType ScriptMethod -Name TestServiceExists -Value { $true } -Force -PassThru
    $Instance.Set()
    It "Does not process if the service is already present" {
      Assert-MockCalled Start-Process -ModuleName TransactSetup -Exactly 0
    }
  }

  Context "Ensure is Absent When Service Exists" {
    Mock Get-Service -ModuleName 'TransactSetup' {$true}

    # Configure context specific properties
    $Instance.Ensure = 'Absent'

    # Run Test
    $Instance.Set()

    It "Passes Correct Installer Path to Start-Process" {
      Assert-MockCalled Start-Process -ModuleName TransactSetup -ParameterFilter {
        $FilePath -eq 'msiexec'
      }
    }
  }

  Context "Ensure is Absent When Serice Does Not Exist" {
    Mock Get-Service -ModuleName TransactSetup {$null}

    # Configure context specific properties
    $Instance = $Instance | Add-Member -MemberType ScriptMethod -Name TestServiceExists -Value { $false } -Force -PassThru
    $Instance.Ensure = 'Absent'

    # Run Test
    $Instance.Set()

    It "Does not process if the service does not exist" {
      Assert-MockCalled Start-Process -ModuleName TransactSetup -Exactly 0
    }
  }

  Context "Version Specified is 2019.2 or earlier" {
    $Version = '2019.2'

    # Context specific Mocks
    Mock Get-Content -ModuleName 'TransactSetup' {'Test=2020'}
    Mock Get-Service -ModuleName 'TransactSetup' {$true}

    # Create new object and do not mock Fix2020CronJobIssue method
    $Params.Version = $Version
    $Params.MethodsToMock = @('PrepareConfigProperties','SetHostWebProperties','SetTransactLicense','StartService','DemandServiceExists')
    $Instance = $Instance = New-TransactSetupObject @Params
    $Instance = $Instance | Add-Member -MemberType ScriptMethod -Name TestServiceExists -Value { $false } -Force -PassThru
    $Instance.Set()

    It "Applies fix to cron job text" {
      Assert-MockCalled Set-Content -ModuleName 'TransactSetup' -ParameterFilter {
           $Value -like "Test=*"
      }
    }
  }

  Context "Version Specified is 2020.1 or later" {
    $Version = '2020.1'

    # Context specific Mocks
    Mock Get-Content -ModuleName 'TransactSetup' {'Test=2020'}
    Mock Get-Service -ModuleName 'TransactSetup' {$true}

    # Create new object and do not mock Fix2020CronJobIssue method
    $Params.Version = $Version
    $Params.MethodsToMock = @('PrepareConfigProperties','SetHostWebProperties','SetTransactLicense','StartService','DemandServiceExists')
    $Instance = $Instance = New-TransactSetupObject @Params
    $Instance = $Instance | Add-Member -MemberType ScriptMethod -Name TestServiceExists -Value { $false } -Force -PassThru
    $Instance.Set()

    It "Does not apply fix to cron job text" {
      Assert-MockCalled Set-Content -ModuleName 'TransactSetup' -Exactly 0
    }
  }
}

examples/TransactSetup.SQL.Example.ps1
Param (
  [Parameter (Mandatory = $true)]
  [string] $IPAddress
)

<#
  Use this example to create a .MOF which will deploy an MS SQL instance using Windows authentication,
  Ephesoft Transact instance, and all prerequisites on a machine via DSC.  All installers will need to
  be pre-staged on the system.  Make sure to update variables as needed.
#>

Configuration SetVMConfig
{
  # Import DSC modules
  Import-DscResource -ModuleName 'ComputerManagementDsc'
  Import-DscResource -ModuleName 'Ephesoft.Transact.Dsc'
  Import-DSCResource -ModuleName 'NetworkingDsc'
  Import-DscResource -ModuleName 'PowerShellModule'
  Import-DscResource -ModuleName 'PSDesiredStateConfiguration'
  Import-DscResource -ModuleName 'SQLServerDSC'
  Import-DscResource -ModuleName 'xPSDesiredStateConfiguration'
  Import-DSCResource -ModuleName 'xSystemSecurity' -Name 'xIEEsc'

  Node $IPAddress {
    LocalConfigurationManager
    {
      RebootNodeIfNeeded = $true
    }

    UserAccountControl 'TransactUAC'
    {
        IsSingleInstance  = 'Yes'
        SuppressRestart   = $true
        NotificationLevel = 'NeverNotifyAndDisableAll'
    }

    Firewall 'TransactTCP'
    {
        Name        = 'TransactTCP'
        DisplayName = 'Transact-8080'
        Action      = 'Allow'
        Direction   = 'Inbound'
        LocalPort   = '8080'
        Protocol    = 'TCP'
        Profile     = 'Any'
        Enabled     = 'True'
    }

    NetAdapterBinding 'DisableIPv6'
    {
        InterfaceAlias = '*'
        ComponentId    = 'ms_tcpip6'
        State          = 'Disabled'
    }

    WindowsFeature 'NetFramework'
    {
         Name   = 'NET-Framework-45-Core'
         Ensure = 'Present'
    }

    # Disable UAC for Administrators only
    xIEEsc 'DisableIEEscAdmin'
    {
        IsEnabled = $False
        UserRole  = "Administrators"
    }

    xIEEsc 'EnableIEEscUser'
    {
        IsEnabled = $True
        UserRole  = "Users"
    }

    xArchive 'UnzipSQLInstaller'
    {
        Path        = 'C:\Path\To\SQL2017.zip'
        Destination = 'C:\Installers\SQLServer\2017'
        Validate    = $true
        Checksum    = 'SHA-256'
        Force       = $true
        Ensure      = 'Present'
    }

    SqlSetup 'InstallDefaultInstance'
    {
      InstanceName        = 'MSSQLSERVER'
      Features            = 'SQLENGINE'
      SourcePath          = 'C:\Installers\SQLServer\2017'
      SQLSysAdminAccounts = @('Administrators')
      DependsOn           = '[WindowsFeature]NetFramework'
      BrowserSvcStartupType = 'Automatic'
      SuppressReboot      = $true
    }

    # These configurations are required as documented here: https://ephesoft.com/docs/microsoft-sql-server-mssql-installation-and-configuration-instructions/
    SqlServerNetwork 'SetTCPConfig'
    {
      InstanceName        = 'MSSQLSERVER'
      ProtocolName        = 'tcp'
      IsEnabled           = $true
      TCPDynamicPort      = $false
      TCPPort             = '1433'
      RestartService      = $true
    }

    Package 'SQLCMDUtils'
    {
        Ensure = 'Present'
        Name = 'Microsoft Command Line Utilities 13 for SQL Server'
        Path = "C:\Installers\SQLServer\MsSqlCmdLnUtils.msi"
        Arguments = 'IACCEPTMSSQLCMDLNUTILSLICENSETERMS=YES'
        ProductId = '33C3E60D-6E22-445C-9B44-E9EEA5C47A01'
        DependsOn = '[SqlSetup]InstallDefaultInstance'
    }

    # Transact
    if ($Transact) {


      TransactSetup 'InstallTransact'
      {
        DatabaseType = 'SQL'
        SQLAuthType = 'Windows'
        Ensure = 'Present'
        InstallerPath = 'C:\Installers\Transact'
        Version = '2020.1'
        LicenseFilePath = 'C:\Installers\License\Ephesoft.lic'
      }

      ServiceSet 'SetTransactServiceToAutomatic'
      {
        Name        = 'EphesoftTransact'
        StartupType = 'Automatic'
        DependsOn   = 'InstallTransact'
      }
    }
  }
}

SetVMConfig -OutputPath 'C:\DSC' -IPAddress $IPAddress -ConfigurationData $ConfigurationData

examples/TransactSetup.MariaDB.Example.ps1
Param (
  [Parameter (Mandatory = $true)]
  [pscredential] $DBCred,
  [Parameter (Mandatory = $true)]
  [string] $IPAddress
)

<#
  This example configuration allows the SQL credential to be stored in the .MOF in plain text.
  Only use this for demo/dev configurations.
  For instructions on securely storing the credential in the .MOF please see https://docs.microsoft.com/en-us/powershell/scripting/dsc/pull-server/securemof?view=powershell-7

  Use this example to create a .MOF which will deploy an MS SQL instance, Ephesoft Transact instance, and all prerequisites on a machine via DSC.

  All installers will need to be pre-staged on the system.  Make sure to update variables as needed.
#>
$ConfigurationData = @{
  AllNodes = @(
    @{
        NodeName=$IPAddress
        PSDscAllowPlainTextPassword=$true
    }
  )
}

Configuration SetConfig
{
  # Import DSC modules
  Import-DscResource -ModuleName 'ComputerManagementDsc'
  Import-DscResource -ModuleName 'Ephesoft.Transact.DSC'
  Import-DSCResource -ModuleName 'NetworkingDsc'
  Import-DscResource -ModuleName 'PowerShellModule'
  Import-DscResource -ModuleName 'PSDesiredStateConfiguration'
  Import-DscResource -ModuleName 'MariaDB.DSC'
  Import-DscResource -ModuleName 'xPSDesiredStateConfiguration'
  Import-DSCResource -ModuleName 'xSystemSecurity' -Name 'xIEEsc'

  Node $IPAddress {
    LocalConfigurationManager
    {
      RebootNodeIfNeeded = $true
    }

    UserAccountControl 'TransactUAC'
    {
        IsSingleInstance  = 'Yes'
        SuppressRestart   = $true
        NotificationLevel = 'NeverNotifyAndDisableAll'
    }

    Firewall 'TransactTCP'
    {
        Name        = 'TransactTCP'
        DisplayName = 'Transact-8080'
        Action      = 'Allow'
        Direction   = 'Inbound'
        LocalPort   = '8080'
        Protocol    = 'TCP'
        Profile     = 'Any'
        Enabled     = 'True'
    }

    NetAdapterBinding 'DisableIPv6'
    {
        InterfaceAlias = '*'
        ComponentId    = 'ms_tcpip6'
        State          = 'Disabled'
    }

    WindowsFeature 'NetFramework'
    {
         Name   = 'NET-Framework-45-Core'
         Ensure = 'Present'
    }

    # Disable UAC for Administrators only
    xIEEsc 'DisableIEEscAdmin'
    {
        IsEnabled = $False
        UserRole  = "Administrators"
    }

    xIEEsc 'EnableIEEscUser'
    {
        IsEnabled = $True
        UserRole  = "Users"
    }

    # MariaDB
    MariaDB 'DBInstance'
    {
      InstallerPath = 'C:\Installers\MariadB\MariaDB.msi'
      RootPwd = $DBCred
      ServiceName = $MariaDBServiceName
      Ensure = $MariaDBEnsure
    }

    # Transact
    if ($Transact) {
      TransactSetup 'InstallTransact'
      {
        DatabaseType = 'MariaDB'
        DBCredential = $DBCred
        Ensure = 'Present'
        InstallerPath = 'C:\Installers\Transact'
        Version = '2020.1'
        LicenseFilePath = 'C:\Installers\License\Ephesoft.lic'
      }

      ServiceSet 'SetTransactServiceToAutomatic'
      {
        Name        = 'EphesoftTransact'
        StartupType = 'Automatic'
        DependsOn   = 'InstallTransact'
      }
    }
  }
}

SetVMConfig -OutputPath 'C:\DSC' -IPAddress $IPAddress -ConfigurationData $ConfigurationData